Added implementation for Splay Trees.

Vanshika266 opened the pull request:
Added code for splay trees with public APIs insert, delete, splay. The work is in reference with issue #109

HarsheetKakar commented:
files have trailing whitespaces in it. If you are using vs code, there is an option to remove trailing whitespaces on save.

#### Reference
https://stackoverflow.com/questions/30884131/remove-trailing-spaces-automatically-or-with-a-shortcut

HarsheetKakar commented:
adding tests will increase coverage. 
https://codecov.io/gh/codezonediitj/pydatastructs/pull/157/diff the lines with red markings dont have tests

czgdp1807 commented:
@Vanshika266 I hope that you are aware that official coding period of RGSoC, 2020 starts from July. Till 31st March you have to submit your application on RGSoC website. See, https://railsgirlssummerofcode.org/students/application/#apply for steps of applying.

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1037,3 +1038,65 @@ def get_sum(self, left_index, right_index):
+        References
+        ==========
+        [1] https://en.wikipedia.org/wiki/Splay_tree
+
```
Please add the following sections to the above doc string.
```
Parameters
==========
...

Examples
========
...
```

czgdp1807 replied:
No need for this. Just checked, it will be repetitive.

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1037,3 +1038,65 @@ def get_sum(self, left_index, right_index):
+                        super(SplayTree,self)._right_rotate(p,x)
+                else:
+                        super(SplayTree,self)._left_rotate(p,x)
+
```
```suggestion
                if self.tree[p].left == x:
                        super(SplayTree, self)._right_rotate(p, x)
                else:
                        super(SplayTree, self)._left_rotate(p, x)
```

Vanshika266 replied:
Changed.

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1037,3 +1038,65 @@ def get_sum(self, left_index, right_index):
+
+        def _zig_zig(self,x,p):
+                super(SplayTree,self)._right_rotate(self.tree[p].parent,p)
+                super(SplayTree,self)._right_rotate(p,x)
```
```suggestion
                super(SplayTree, self)._right_rotate(self.tree[p].parent, p)
                super(SplayTree, self)._right_rotate(p, x)
```

Vanshika266 replied:
Fixed.

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1037,3 +1038,65 @@ def get_sum(self, left_index, right_index):
+
+        def _zig_zag(self,x,p):
+                super(SplayTree,self)._left_rotate(p,x)
+                super(SplayTree,self)._right_rotate(self.tree[x].parent,x)
```
```suggestion
                super(SplayTree, self)._left_rotate(p, x)
                super(SplayTree, self)._right_rotate(self.tree[x].parent, x)
```

Vanshika266 replied:
Changed.

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1037,3 +1038,65 @@ def get_sum(self, left_index, right_index):
+                super(SplayTree,self)._right_rotate(self.tree[x].parent,x)
+
+        def _zag_zag(self,x,p):
+                super(SplayTree,self)._left_rotate(self.tree[p].parent,p)
```
```suggestion
                super(SplayTree, self)._left_rotate(self.tree[p].parent, p)
```

Vanshika266 replied:
Changed.

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1037,3 +1038,65 @@ def get_sum(self, left_index, right_index):
+
+        def _zag_zag(self,x,p):
+                super(SplayTree,self)._left_rotate(self.tree[p].parent,p)
+                super(SplayTree,self)._left_rotate(p,x)
```
```suggestion
                super(SplayTree, self)._left_rotate(p, x)
```

Vanshika266 replied:
Fixed.

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1037,3 +1038,65 @@ def get_sum(self, left_index, right_index):
+
+        def _zag_zig(self,x,p):
+                super(SplayTree,self)._right_rotate(p,x)
+                super(SplayTree,self)._left_rotate(self.tree[x].parent,x)
```
```suggestion
                super(SplayTree, self)._right_rotate(p, x)
                super(SplayTree, self)._left_rotate(self.tree[x].parent, x)
```

Vanshika266 replied:
Changed.

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1037,3 +1038,65 @@ def get_sum(self, left_index, right_index):
+                                self._zag_zig(x,p)
+                        else:
+                                self._zig_zag(x,p)
+                        p = self.tree[x].parent
```
```suggestion
                while self.tree[x].parent is not None:
                        if self.tree[p].parent is None:
                                self._zig(x, p)
                        elif self.tree[p].left == x and self.tree[self.tree[p].parent].left == p):
                                self._zig_zig(x, p)
                        elif self.tree[p].right == x and self.tree[self.tree[p].parent].right == p:
                                self._zag_zag(x, p)
                        elif self.tree[p].left == x and self.tree[self.tree[p].parent].right == p:
                                self._zag_zig(x, p)
                        else:
                                self._zig_zag(x, p)
                        p = self.tree[x].parent
```

Vanshika266 replied:
Changed.

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1037,3 +1038,65 @@ def get_sum(self, left_index, right_index):
+                        else:
+                                self._zig_zag(x,p)
+                        p = self.tree[x].parent
+
```
```suggestion
```

Vanshika266 replied:
Fixed.

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1037,3 +1038,65 @@ def get_sum(self, left_index, right_index):
+
+        def insert(self,key,x):
+                super(AVLTree, self).insert(key, x)
+                e, p = super(AVLTree,self).search(x,parent=True)
```
```suggestion
                e, p = super(AVLTree, self).search(x, parent=True)
```

Vanshika266 replied:
Fixed.

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1037,3 +1038,65 @@ def get_sum(self, left_index, right_index):
+
+        def delete(self,x):
+                b = super(AVLTree,self).delete(x,balancing_info=True)
+                self.splay(b,self.tree[b].parent)
```
```suggestion
                b = super(AVLTree, self).delete(x, balancing_info=True)
                self.splay(b, self.tree[b].parent)
```

Vanshika266 replied:
Changed.

Vanshika266 commented:
> @Vanshika266 I hope that you are aware that official coding period of RGSoC, 2020 starts from July. Till 31st March you have to submit your application on RGSoC website. See, https://railsgirlssummerofcode.org/students/application/#apply for steps of applying.

Yes, I will start working on the application.

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1037,3 +1038,64 @@ def get_sum(self, left_index, right_index):
+
+class SplayTree(AVLTree):
+        """
+        Represents Splay Trees.
```
```suggestion
        Represents Splay Trees.

```

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1037,3 +1038,65 @@ def get_sum(self, left_index, right_index):
+                                self._zig_zag(x, p)
+                        p = self.tree[x].parent
+
+        def insert(self,key,x):
```
```suggestion
        def insert(self, key, x):
```

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1037,3 +1038,65 @@ def get_sum(self, left_index, right_index):
+                super(SplayTree, self)._left_rotate(self.tree[p].parent, p)
+                super(SplayTree, self)._left_rotate(p, x)
+
+        def _zag_zig(self,x,p):
```
```suggestion
        def _zig(self, x, p):
                if self.tree[p].left == x:
                        super(SplayTree, self)._right_rotate(p, x)
                else:
                        super(SplayTree, self)._left_rotate(p, x)

        def _zig_zig(self, x, p):
                super(SplayTree, self)._right_rotate(self.tree[p].parent, p)
                super(SplayTree, self)._right_rotate(p, x)

        def _zig_zag(self, x, p):
                super(SplayTree, self)._left_rotate(p, x)
                super(SplayTree, self)._right_rotate(self.tree[x].parent, x)

        def _zag_zag(self, x, p):
                super(SplayTree, self)._left_rotate(self.tree[p].parent, p)
                super(SplayTree, self)._left_rotate(p, x)

        def _zag_zig(self, x, p):
```

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1037,3 +1038,65 @@ def get_sum(self, left_index, right_index):
+                raise NotImplementedError("This is not implemented yet.")
+
+        def split(self,x):
+                raise NotImplementedError("This is not implemented yet.")
```
```suggestion

class SplayTree(SelfBalancingBinaryTree):
        """
        Represents Splay Trees.

        References
        ==========
        [1] https://en.wikipedia.org/wiki/Splay_tree

        """
        def _zig(self, x, p):
                if self.tree[p].left == x:
                        super(SplayTree, self)._right_rotate(p, x)
                else:
                        super(SplayTree, self)._left_rotate(p, x)

        def _zig_zig(self, x, p):
                super(SplayTree, self)._right_rotate(self.tree[p].parent, p)
                super(SplayTree, self)._right_rotate(p, x)

        def _zig_zag(self, x, p):
                super(SplayTree, self)._left_rotate(p, x)
                super(SplayTree, self)._right_rotate(self.tree[x].parent, x)

        def _zag_zag(self, x, p):
                super(SplayTree, self)._left_rotate(self.tree[p].parent, p)
                super(SplayTree, self)._left_rotate(p, x)

        def _zag_zig(self, x, p):
                super(SplayTree, self)._right_rotate(p, x)
                super(SplayTree, self)._left_rotate(self.tree[x].parent, x)

        def splay(self,x,p):
                while self.tree[x].parent is not None:
                        if self.tree[p].parent is None:
                                self._zig(x, p)
                        elif self.tree[p].left == x and self.tree[self.tree[p].parent].left == p:
                                self._zig_zig(x, p)
                        elif self.tree[p].right == x and self.tree[self.tree[p].parent].right == p:
                                self._zag_zag(x, p)
                        elif self.tree[p].left == x and self.tree[self.tree[p].parent].right == p:
                                self._zag_zig(x, p)
                        else:
                                self._zig_zag(x, p)
                        p = self.tree[x].parent

        def insert(self, key, x):
                super(SelfBalancingBinaryTree, self).insert(key, x)
                e, p = super(SelfBalancingBinaryTree, self).search(x, parent=True)
                self.tree[self.size-1].parent = p;
                self.splay(e, p)

        def delete(self,x):
                b = super(SelfBalancingBinaryTree, self).delete(x, balancing_info=True)
                self.splay(b, self.tree[b].parent)
                return True

        def join(self,tree2):
                raise NotImplementedError("This is not implemented yet.")

        def split(self,x):
                raise NotImplementedError("This is not implemented yet.")
```

Vanshika266 replied:
Updated.

czgdp1807 commented:
@Vanshika266 Before adding implementation for `join` and `split` methods, I'd suggest to test whether the changes made in #176 works correctly or not. Please add tests for `insert` and `delete`. You can take examples from wikipedia and add tests on the basis of those examples.

czgdp1807 commented:
@Vanshika266 
**Reminder**
I hope that your team is able to find coaches for completing the application process for RGSoC, 2020. If not then please feel free to send invites to [me](https://teams.railsgirlssummerofcode.org/users/4331) and [Vedansh Priyadarshi](https://teams.railsgirlssummerofcode.org/users/4421) as the deadline for application is quite near.

Vanshika266 commented:
> @Vanshika266 Before adding implementation for `join` and `split` methods, I'd suggest to test whether the changes made in #176 works correctly or not. Please add tests for `insert` and `delete`. You can take examples from wikipedia and add tests on the basis of those examples.

I have added tests for `insert`, `delete`, `split` and `join` methods in SplayTree.
Can you review the tests once to see if it works ?

czgdp1807 commented:
Coverage is low. More tests are needed for the lines marked in red in [this page](https://codecov.io/gh/codezonediitj/pydatastructs/pull/157/diff).

czgdp1807 commented:
I am looking into the coverage issue. I will let you know if things out work well. If the coverage issue is fixed this is ready to merge.

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1053,3 +1054,85 @@ def get_sum(self, left_index, right_index):
+        Represents Splay Trees.
+        References
+        ==========
+        [1] https://en.wikipedia.org/wiki/Splay_tree
```
```suggestion
        .. [1] https://en.wikipedia.org/wiki/Splay_tree
```

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1053,3 +1054,85 @@ def get_sum(self, left_index, right_index):
+        """
+        Represents Splay Trees.
+        References
+        ==========
```
```suggestion
        ==========

```

czgdp1807 commented:
I observed that there is a bug in the code,
```python
>>> from pydatastructs import SplayTree
>>> t = SplayTree(100, 100)
>>> t.insert(50, 50)
>>> t.insert(200, 200)
>>> t.insert(40, 40)
>>> t.insert(30, 30)
>>> t.insert(20, 20)
>>> t.insert(55, 55)
>>> str(t)
'[(None, 100, 100, None), (6, 50, 50, None), (3, 200, 200, 1), (None, 40, 40, 1), (5, 30, 30, 2), (None, 20, 20, None), (4, 55, 55, 0)]'
```
The tree constructed is a bit wrong. Given below is the visualisation,
![f](https://user-images.githubusercontent.com/36567889/77791603-f6688a00-708c-11ea-93c9-2f23c4c01406.png)

You can visualise your trees at https://www.cs.usfca.edu/~galles/visualization/SplayTree.html and compare the results for correctness.

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1053,3 +1054,86 @@ def get_sum(self, left_index, right_index):
+                    self.tree[self.root_idx].right = None
+                return other
+
+        def split_array(self, sec, x):
```
Please add docstring for this method as this isn't common among various Tree classes.

Vanshika266 replied:
Done.

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1053,3 +1054,86 @@ def get_sum(self, left_index, right_index):
+                self.splay(b, self.tree[b].parent)
+                return True
+
+        def join(self, other):
```
+1

Vanshika266 replied:
Done.

Vanshika266 replied:
Added docstring for `join` method.

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1053,3 +1054,86 @@ def get_sum(self, left_index, right_index):
+                self.splay(maxm, self.tree[maxm].parent)
+                other.split_array(self, other.tree[other.root_idx].data)
+
+        def split(self, x):
```
+1

Vanshika266 replied:
Added docstring for `split` method.

Vanshika266 commented:
> I observed that there is a bug in the code,
> 
> ```python
> >>> from pydatastructs import SplayTree
> >>> t = SplayTree(100, 100)
> >>> t.insert(50, 50)
> >>> t.insert(200, 200)
> >>> t.insert(40, 40)
> >>> t.insert(30, 30)
> >>> t.insert(20, 20)
> >>> t.insert(55, 55)
> >>> str(t)
> '[(None, 100, 100, None), (6, 50, 50, None), (3, 200, 200, 1), (None, 40, 40, 1), (5, 30, 30, 2), (None, 20, 20, None), (4, 55, 55, 0)]'
> ```
> 
> The tree constructed is a bit wrong. Given below is the visualisation for the correct structure,
> ![f](https://user-images.githubusercontent.com/36567889/77791603-f6688a00-708c-11ea-93c9-2f23c4c01406.png)
> 
> You can visualise your trees at https://www.cs.usfca.edu/~galles/visualization/SplayTree.html and compare the results for correctness.

I am sorry, I found the bug in the code and have fixed it. Also, I have updated the code for `delete` method and followed the implementation variant given on https://en.wikipedia.org/wiki/Splay_tree. Can you review it once to see if it is the required implementation ?

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1053,3 +1054,131 @@ def get_sum(self, left_index, right_index):
+
+class SplayTree(SelfBalancingBinaryTree):
+    """
+    Represents Splay Trees.
```
```suggestion
    Represents Splay Trees.

```

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1053,3 +1054,131 @@ def get_sum(self, left_index, right_index):
+    Represents Splay Trees.
+    References
+    ==========
+    [1] https://en.wikipedia.org/wiki/Splay_tree
```
```suggestion
    .. [1] https://en.wikipedia.org/wiki/Splay_tree
```

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1053,3 +1054,131 @@ def get_sum(self, left_index, right_index):
+    def _zag_zig(self,x,p):
+        super(SplayTree, self)._right_left_rotate(self.tree[p].parent, p)
+
+    def splay(self,x,p):
```
```suggestion
    def splay(self, x, p):
```

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1053,3 +1054,131 @@ def get_sum(self, left_index, right_index):
+    [1] https://en.wikipedia.org/wiki/Splay_tree
+
+    """
+    def _zig(self,x,p):
```
```suggestion
    def _zig(self, x, p):
```

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1053,3 +1054,131 @@ def get_sum(self, left_index, right_index):
+        else:
+            super(SplayTree, self)._left_rotate(p, x)
+
+    def _zig_zig(self,x,p):
```
```suggestion
    def _zig_zig(self, x, p):
```

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1053,3 +1054,131 @@ def get_sum(self, left_index, right_index):
+        super(SplayTree, self)._right_rotate(self.tree[p].parent, p)
+        super(SplayTree, self)._right_rotate(p, x)
+
+    def _zig_zag(self,x,p):
```
```suggestion
    def _zig_zag(self, x, p):
```

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1053,3 +1054,131 @@ def get_sum(self, left_index, right_index):
+    def _zig_zag(self,x,p):
+        super(SplayTree, self)._left_right_rotate(self.tree[p].parent, p)
+
+    def _zag_zag(self,x,p):
```
```suggestion
    def _zag_zag(self, x, p):
```

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1053,3 +1054,131 @@ def get_sum(self, left_index, right_index):
+        super(SplayTree, self)._left_rotate(self.tree[p].parent, p)
+        super(SplayTree, self)._left_rotate(p, x)
+
+    def _zag_zig(self,x,p):
```
```suggestion
    def _zag_zig(self, x, p):
```

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1053,3 +1054,131 @@ def get_sum(self, left_index, right_index):
+                self._zig_zag(x, p)
+            p = self.tree[x].parent
+
+    def insert(self,key,x):
```
```suggestion
    def insert(self, key, x):
```

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1053,3 +1054,131 @@ def get_sum(self, left_index, right_index):
+        self.tree[self.size-1].parent = p;
+        self.splay(e, p)
+
+    def delete(self,x):
```
```suggestion
    def delete(self, x):
```

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1053,3 +1054,132 @@ def get_sum(self, left_index, right_index):
+
+    def split_array(self, sec, x):
+        """
+        Helper function that appends all elements of the SplayTree self into
```
```suggestion
        Helper function that appends all elements of the current splay tree into
```

Vanshika266 replied:
Done.

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1053,3 +1054,132 @@ def get_sum(self, left_index, right_index):
+
+    def split(self, x):
+        """
+        Splits self tree into two trees such that one tree contains nodes
```
```suggestion
        Splits current splay tree into two trees such that one tree contains nodes
```

Vanshika266 replied:
Done.

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1053,3 +1054,132 @@ def get_sum(self, left_index, right_index):
+        ==========
+
+        sec: SplayTree
+            Elements of self tree are appended in sec.
```
```suggestion
            Elements of current tree are appended in sec.
```

Vanshika266 replied:
Done.

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1053,3 +1054,132 @@ def get_sum(self, left_index, right_index):
+            self.split_array(sec, l.data)
+        if self.tree[e].right is not None:
+            r = self.tree[self.tree[e].right]
+            self.split_array(sec, r.data)
```
Can you convert this into iterative algorithm using stacks?

Vanshika266 replied:
Done.

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1053,3 +1054,132 @@ def get_sum(self, left_index, right_index):
+            self.tree[self.root_idx].right = None
+        return other
+
+    def split_array(self, sec, x):
```
Should this be public?

Vanshika266 replied:
Removed split_array function, as it was no longer useful in iterative algorithm using stacks.

czgdp1807 commented:
Any updates?

Vanshika266 commented:
Sorry, pushed these commits by mistake. Fixing it.

Vanshika266 commented:
> Any updates?

Sorry for the delay. I have updated the implementation of SplayTree. I have removed the function split_array, and have now used the implementation of pre-order traversal from BinaryTreeTraversal Class. 
Can you review the implementation once ?

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1059,3 +1060,123 @@ def get_sum(self, left_index, right_index):
+        elements = traverse.depth_first_search(order='pre_order', node=other.root_idx)
+        for i in range(len(elements)):
+            super(SelfBalancingBinaryTree, self).insert(elements[i].key, elements[i].data)
+        j = len(elements)-1
+        while j>=0:
```
```suggestion
        for j in range(len(elements) - 1, -1, -1):
```

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1059,3 +1060,123 @@ def get_sum(self, left_index, right_index):
+        while j>=0:
+            e, p = super(SelfBalancingBinaryTree, other).search(elements[j].key, parent=True)
+            other.tree[e] = None
+            j-=1
```
```suggestion
```

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1059,3 +1060,123 @@ def get_sum(self, left_index, right_index):
+            elements = traverse.depth_first_search(order='pre_order', node=self.tree[self.root_idx].right)
+            for i in range(len(elements)):
+                super(SelfBalancingBinaryTree, other).insert(elements[i].key, elements[i].data)
+            j = len(elements)-1
+            while j>=0:
```
```suggestion
            for j in range(len(elements) - 1, -1, -1):
```

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1059,3 +1060,123 @@ def get_sum(self, left_index, right_index):
+            while j>=0:
+                e, p = super(SelfBalancingBinaryTree, self).search(elements[j].key, parent=True)
+                self.tree[e] = None
+                j-=1
```
```suggestion
```

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1059,3 +1060,123 @@ def get_sum(self, left_index, right_index):
+    Represents Splay Trees.
+
+    References
+    ==========
```
```suggestion
    ==========

```

czgdp1807 commented on pydatastructs/trees/tests/test_binary_trees.py:
```diff
@@ -348,3 +348,29 @@ def test_issue_234():
+
+    assert str(t) == ("[(None, 100, 100, None), (None, 50, 50, None), (0, 200, 200, None), (None, 40, 40, 1), (5, 30, 30, 3), (None, 20, 20, None), (4, 55, 55, 2)]")
+    t.delete(40)
+    assert str(t) == ("[(None, 100, 100, None), '', (0, 200, 200, None), (4, 50, 50, 6), (5, 30, 30, None), (None, 20, 20, None), (None, 55, 55, 2)]")
```
![tree](https://user-images.githubusercontent.com/36567889/78768593-c92fac00-79a9-11ea-977a-2b57d1442cad.png)
I think this is a bug.

czgdp1807 replied:
It's not a bug. It's just a difference in methodologies. 
You have used the following,
> To delete a node x, use the same method as with a binary search tree:

> If x has two children:
Swap its value with that of either the rightmost node of its left sub tree (its in-order predecessor) or the leftmost node of its right subtree (its in-order successor).
Remove that node instead.
In this way, deletion is reduced to the problem of removing a node with 0 or 1 children. Unlike a binary search tree, in a splay tree after deletion, we splay the parent of the removed node to the top of the tree.


The website has used the following,

> Alternatively:

> The node to be deleted is first splayed, i.e. brought to the root of the tree and then deleted. leaves the tree with two sub trees.
The two sub-trees are then joined using a "join" operation.

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -1059,3 +1060,120 @@ def get_sum(self, left_index, right_index):
+        super(SplayTree, self)._left_rotate(self.tree[p].parent, p)
+        super(SplayTree, self)._left_rotate(p, x)
+
+    def _zag_zig(self, x, p):
```
`x` isn't needed here, IMO.

Vanshika266 replied:
Yes, I realized `x` can be removed. I have updated the implementation. Please verify.

czgdp1807 commented:
@Vanshika266 Can you please enable `Allow edits from maintainers`(on the right side under labels)?

czgdp1807 commented on pydatastructs/trees/binary_trees.py:
```diff
@@ -278,6 +280,7 @@ def delete(self, key, **kwargs):
 
         elif self.tree[walk].left is not None and \
             self.tree[walk].right is not None:
+            print(walk)
```
```suggestion
```

czgdp1807 commented:
Thanks.
